File: app/src/main/java/workshop/akbolatss/xchangesrates/screens/charts/ChartPresenter.java
package workshop.akbolatss.xchangesrates.screens.charts;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableSingleObserver;
import io.reactivex.schedulers.Schedulers;
import workshop.akbolatss.xchangesrates.base.BasePresenter;
import workshop.akbolatss.xchangesrates.model.ExchangeModel;
import workshop.akbolatss.xchangesrates.model.mapper.ChartDataMapper;
import workshop.akbolatss.xchangesrates.model.response.ChartResponse;
import workshop.akbolatss.xchangesrates.model.response.ChartResponseData;
import workshop.akbolatss.xchangesrates.repositories.DBChartRepository;

/**
 * Author: Akbolat Sadvakassov
 * Date: 04.01.2018
 */

public class ChartPresenter extends BasePresenter<ChartView> {

    /**
     * Репозитории
     */
    private DBChartRepository mRepository;

    /**
     * Main model
     */
    private ExchangeModel mExchangeModel;

    /**
     * Response от сервера
     */
    private ChartResponseData mChartData;

    private String mCoinCode;
    private String mCurrencyCode;
    private String mTerm;

    public ChartPresenter(DBChartRepository mRepository) {
        this.mRepository = mRepository;
    }

    private CompositeDisposable compositeDisposable;

    @Override
    public void onViewAttached(ChartView view) {
        super.onViewAttached(view);
        compositeDisposable = new CompositeDisposable();
    }

    @Override
    public void onViewDetached(ChartView view) {
        super.onViewDetached(view);
        if (compositeDisposable != null) {
            compositeDisposable.clear();
        }
    }

    public void onUpdate() {
        getView().onShowLoading();
        compositeDisposable.add(
                mRepository.getQueryResult(mCoinCode,
                        mExchangeModel.getIdent(),
                        mCurrencyCode,
                        mTerm)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(new DisposableSingleObserver<ChartResponse>() {
                            @Override
                            public void onSuccess(ChartResponse response) {
                                mChartData = response.getData();
                                mChartData.setCoin(mCoinCode);
                                getView().onLoadLineChart(mChartData);
                                getView().onHideLoading();
                            }

                            @Override
                            public void onError(Throwable e) {
                                getView().onHideLoading();
                            }
                        }));
    }

    public void onSaveSnap() {
        getView().onShowLoading();
        ChartDataMapper mapper = new ChartDataMapper(mChartData, mChartData.getInfo(), mChartData.getChart());
//        Flowable<Object> s = Single.concat(mapper.transformD(mChartData), //TODO Улучши, когда твои скиллы по RxJava 2 поднимутся
//                mapper.transformInfoD(mChartData.getInfo()),
//                mapper.transformChartsD(mChartData.getChart()));
        compositeDisposable.add(
                mRepository.onAddChartData(mapper.getData(), mapper.getDataInfo(), mapper.getChartsList())
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(new DisposableSingleObserver<Boolean>() {
                            @Override
                            public void onSuccess(Boolean aBoolean) {
                                getView().onHideLoading();
                            }

                            @Override
                            public void onError(Throwable e) {
                                getView().onHideLoading();
                            }
                        }));
    }

    public void setExchangeModel(ExchangeModel mExchangeModel) {
        this.mExchangeModel = mExchangeModel;
    }

    public ExchangeModel getExchangeModel() {
        return mExchangeModel;
    }

    public void setCoinCode(String mCoinCode) {
        this.mCoinCode = mCoinCode;
    }

    public void setCurrencyCode(String mCurrencyCode) {
        this.mCurrencyCode = mCurrencyCode;
    }

    public void setTerm(String mTerm) {
        this.mTerm = mTerm;
    }
}

File: app/src/main/java/workshop/akbolatss/xchangesrates/screens/charts/ChartPresenter.kt
package workshop.akbolatss.xchangesrates.screens.charts

import android.util.Log
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import workshop.akbolatss.xchangesrates.base.BasePresenter
import workshop.akbolatss.xchangesrates.model.ExchangeModel
import workshop.akbolatss.xchangesrates.model.response.ChartData
import workshop.akbolatss.xchangesrates.model.response.ChartOptions
import workshop.akbolatss.xchangesrates.repositories.DBChartRepository
import workshop.akbolatss.xchangesrates.utils.Constants
import java.util.ArrayList

class ChartPresenter(private val mRepository: DBChartRepository) : BasePresenter<ChartView>() {

    /**
     * Main model
     */
    lateinit var exchangeModel: ExchangeModel
    var exchangeModelPos: Int = 0
    /**
     * Chart Model
     */
    lateinit var mChartData: ChartData
    /**
     * Timing, e.g. 10min, 1h, 3h, 12h, 1 day, etc.
     */
    lateinit var mTerm: String
    /**
     * ID of crypto valency, btc, eth...
     */
    lateinit var mCoinCode: String
    var coinCodePos: Int = 0
    /**
     * ID of standard valency, KZT, USD...
     */
    lateinit var mCurrencyCode: String
    var currencyCodePos: Int = 0

    private var compositeDisposable: CompositeDisposable? = null

    override fun onViewAttached(view: ChartView?) {
        super.onViewAttached(view)
        compositeDisposable = CompositeDisposable()
    }

    override fun onViewDetached(view: ChartView?) {
        super.onViewDetached(view)
        if (compositeDisposable != null) {
            compositeDisposable!!.clear()
        }
    }

    fun onUpdate() {
//        view.onShowLoading()
        compositeDisposable!!.add(
                mRepository.getQueryResult(mCoinCode,
                        exchangeModel.ident,
                        mCurrencyCode,
                        mTerm)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe({ response ->
                            mChartData = response.data
                            mChartData.coin = mCoinCode
                            view.onLoadLineChart(mChartData)
//                            view.onHideLoading()
                        }, {
                            //                            view.onHideLoading()
                        }))
    }

    fun saveChartData() {
        mChartData.coin = mCoinCode
        mChartData.currency = mCurrencyCode
        mChartData.timingName = "12h"
        mChartData.timingIndex = 3 // 12 hours
        mChartData.options = ChartOptions()
        mChartData.options.isSmartEnabled = true
        mChartData.isNotificationEnabled = false
        compositeDisposable!!.add(
                mRepository.onAddChartData(mChartData)
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeOn(Schedulers.io())
                        .subscribe({
                            view.toast("Successfully saved!")
                            mChartData.id = it
                            view.onSaveSnapshot(true, mChartData)
                        }, {
                            view.onSaveSnapshot(false, mChartData)
                            it.printStackTrace()
                            Log.e("TAG", "Error: ${it.message}")
                            view.toast("Failed saving! >(")
                        })
        )
    }


    fun generateButtons(): ArrayList<String> {
        val strings = ArrayList<String>()
        strings.add(Constants.MINUTES_10)
        strings.add(Constants.HOUR_1)
        strings.add(Constants.HOUR_3)
        strings.add(Constants.HOUR_12)
        strings.add(Constants.HOUR_24)
        strings.add(Constants.WEEK)
        strings.add(Constants.MONTH)
        strings.add(Constants.MONTH_3)
        strings.add(Constants.MONTH_6)
        strings.add(Constants.YEAR_1)
        strings.add(Constants.YEAR_2)
        strings.add(Constants.YEAR_5)
        return strings
    }
}
